File: src/services/metrics.service.ts
interface SystemMetrics {
  timestamp: string;
  cpu_usage: number;
  memory_usage: number;
  disk_usage: number;
  active_cameras: number;
}

interface CameraMetrics {
  camera_id: string;
  fps_actual: number;
  fps_target: number;
  latency_ms: number;
  status: string;
  last_detection?: string;
}

interface DetectionMetrics {
  camera_id: string;
  confidence: number;
  label: string;
  is_shoplifting: boolean;
  timestamp: string;
  alert_triggered: boolean;
}

interface MetricsSummary {
  system: SystemMetrics;
  cameras: CameraMetrics[];
  recent_detections: DetectionMetrics[];
  total_detections_today: number;
  alert_count_today: number;
}

interface CameraPerformance {
  camera_id: string;
  time_range: string;
  data: Array<{
    timestamp: string;
    fps_actual: number;
    latency_ms: number;
    queue_depth: number;
    dropped_frames: number;
  }>;
}

interface Alert {
  timestamp: string;
  camera_id: string;
  level: string;
  type: string;
  confidence: number;
  label: string;
}

interface HealthStatus {
  elasticsearch: boolean;
  prometheus: boolean;
  system_monitor: boolean;
  timestamp: string;
}

class MetricsService {
  private baseUrl: string;

  constructor(baseUrl: string = '/api/metrics') {
    this.baseUrl = baseUrl;
  }

  /**
   * Get system performance metrics over time
   */
  async getSystemMetrics(timeRange: string = '15m', limit: number = 100): Promise<SystemMetrics[]> {
    const response = await fetch(`${this.baseUrl}/system?time_range=${timeRange}&limit=${limit}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch system metrics: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get current metrics for all cameras
   */
  async getCameraMetrics(): Promise<CameraMetrics[]> {
    const response = await fetch(`${this.baseUrl}/cameras`);
    if (!response.ok) {
      throw new Error(`Failed to fetch camera metrics: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get detailed performance metrics for a specific camera
   */
  async getCameraPerformance(cameraId: string, timeRange: string = '1h'): Promise<CameraPerformance> {
    const response = await fetch(`${this.baseUrl}/cameras/${cameraId}/performance?time_range=${timeRange}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch camera ${cameraId} performance: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get detection metrics with optional filtering
   */
  async getDetectionMetrics(
    timeRange: string = '1h',
    cameraId?: string,
    confidenceThreshold: number = 0.0
  ): Promise<DetectionMetrics[]> {
    const params = new URLSearchParams({
      time_range: timeRange,
      confidence_threshold: confidenceThreshold.toString()
    });

    if (cameraId) {
      params.append('camera_id', cameraId);
    }

    const response = await fetch(`${this.baseUrl}/detections?${params}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch detection metrics: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get comprehensive metrics summary
   */
  async getMetricsSummary(): Promise<MetricsSummary> {
    const response = await fetch(`${this.baseUrl}/summary`);
    if (!response.ok) {
      throw new Error(`Failed to fetch metrics summary: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get health status of metrics infrastructure
   */
  async getHealthStatus(): Promise<HealthStatus> {
    const response = await fetch(`${this.baseUrl}/health`);
    if (!response.ok) {
      throw new Error(`Failed to fetch health status: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get recent alerts from the system
   */
  async getRecentAlerts(limit: number = 50, severity?: string): Promise<Alert[]> {
    const params = new URLSearchParams({ limit: limit.toString() });
    if (severity) {
      params.append('severity', severity);
    }

    const response = await fetch(`${this.baseUrl}/alerts/recent?${params}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch recent alerts: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Create WebSocket connection for real-time metrics updates
   */
  createMetricsWebSocket(): WebSocket {
    const { protocol, hostname } = window.location;
    const wsProtocol = protocol === 'https:' ? 'wss' : 'ws';
    const wsPort = process.env.NODE_ENV === 'development' ? '8001' : (window.location.port || '8001');
    return new WebSocket(`${wsProtocol}://${hostname}:${wsPort}/ws/metrics`);
  }

  /**
   * Create WebSocket connection for camera-specific metrics
   */
  createCameraMetricsWebSocket(cameraId: string): WebSocket {
    const { protocol, hostname } = window.location;
    const wsProtocol = protocol === 'https:' ? 'wss' : 'ws';
    const wsPort = process.env.NODE_ENV === 'development' ? '8001' : (window.location.port || '8001');
    return new WebSocket(`${wsProtocol}://${hostname}:${wsPort}/ws/metrics/camera/${cameraId}`);
  }

  /**
   * Create WebSocket connection for real-time alerts
   */
  createAlertsWebSocket(): WebSocket {
    const { protocol, hostname } = window.location;
    const wsProtocol = protocol === 'https:' ? 'wss' : 'ws';
    const wsPort = process.env.NODE_ENV === 'development' ? '8001' : (window.location.port || '8001');
    return new WebSocket(`${wsProtocol}://${hostname}:${wsPort}/ws/alerts`);
  }

  /**
   * Parse WebSocket message and return typed data
   */
  parseWebSocketMessage(message: MessageEvent): any {
    try {
      return JSON.parse(message.data);
    } catch (error) {
      console.error('Failed to parse WebSocket message:', error);
      return null;
    }
  }

  /**
   * Get analytics data for a specific time range
   */
  async getAnalyticsData(timeRange: string = '24h'): Promise<{
    systemMetrics: SystemMetrics[];
    detectionMetrics: DetectionMetrics[];
    summary: MetricsSummary;
  }> {
    const [systemMetrics, detectionMetrics, summary] = await Promise.all([
      this.getSystemMetrics(timeRange, 100),
      this.getDetectionMetrics(timeRange),
      this.getMetricsSummary()
    ]);

    return { systemMetrics, detectionMetrics, summary };
  }

  /**
   * Get detection statistics for analytics
   */
  async getDetectionStatistics(timeRange: string = '24h'): Promise<{
    totalDetections: number;
    shopliflingDetections: number;
    averageConfidence: number;
    detectionsByHour: Array<{ hour: string; count: number }>;
  }> {
    const detections = await this.getDetectionMetrics(timeRange);

    const totalDetections = detections.length;
    const shopliflingDetections = detections.filter(d => d.is_shoplifting).length;
    const averageConfidence = detections.length > 0
      ? detections.reduce((sum, d) => sum + d.confidence, 0) / detections.length
      : 0;

    // Group detections by hour
    const detectionsByHour = this.groupDetectionsByHour(detections);

    return {
      totalDetections,
      shopliflingDetections,
      averageConfidence,
      detectionsByHour
    };
  }

  /**
   * Group detections by hour for time-based analytics
   */
  private groupDetectionsByHour(detections: DetectionMetrics[]): Array<{ hour: string; count: number }> {
    const hourCounts: { [key: string]: number } = {};

    detections.forEach(detection => {
      const hour = new Date(detection.timestamp).getHours();
      const hourKey = `${hour}:00`;
      hourCounts[hourKey] = (hourCounts[hourKey] || 0) + 1;
    });

    // Create array for all 24 hours
    const result = [];
    for (let i = 0; i < 24; i++) {
      const hourKey = `${i}:00`;
      result.push({
        hour: hourKey,
        count: hourCounts[hourKey] || 0
      });
    }

    return result;
  }

  /**
   * Format timestamp for display
   */
  formatTimestamp(timestamp: string): string {
    return new Date(timestamp).toLocaleString();
  }

  /**
   * Calculate percentage with proper formatting
   */
  formatPercentage(value: number, decimals: number = 1): string {
    return `${value.toFixed(decimals)}%`;
  }

  /**
   * Get status color for UI elements
   */
  getStatusColor(status: string): string {
    switch (status.toLowerCase()) {
      case 'online':
      case 'active':
      case 'healthy':
        return '#22c55e'; // green-500
      case 'offline':
      case 'inactive':
      case 'error':
        return '#ef4444'; // red-500
      case 'warning':
      case 'degraded':
        return '#f59e0b'; // amber-500
      default:
        return '#6b7280'; // gray-500
    }
  }

  /**
   * Get confidence level label
   */
  getConfidenceLevel(confidence: number): string {
    if (confidence >= 0.9) return 'critical';
    if (confidence >= 0.7) return 'high';
    if (confidence >= 0.5) return 'medium';
    return 'low';
  }

  /**
   * Get confidence level color
   */
  getConfidenceLevelColor(confidence: number): string {
    if (confidence >= 0.9) return '#dc2626'; // red-600
    if (confidence >= 0.7) return '#ea580c'; // orange-600
    if (confidence >= 0.5) return '#d97706'; // amber-600
    return '#65a30d'; // lime-600
  }
}

// Export singleton instance
export const metricsService = new MetricsService();

// Export types for use in components
export type {
  SystemMetrics,
  CameraMetrics,
  DetectionMetrics,
  MetricsSummary,
  CameraPerformance,
  Alert,
  HealthStatus
};

File: src/services/metrics_service.py
import asyncio
import json
import os
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Union

import aiohttp
import psutil
from loguru import logger

from src.utils.system_monitor import SystemMonitor

# Import monitoring with error handling
try:
    from utils.monitoring import SystemMonitor as PrometheusMonitor
    from utils.monitoring import get_monitor
except ImportError as e:
    logger.warning(f"Failed to import Prometheus monitoring: {e}")
    PrometheusMonitor = None
    get_monitor = None


class MetricsService:
    """Unified service for aggregating metrics from multiple sources."""

    def __init__(self):
        self.elasticsearch_url = os.getenv("ELASTICSEARCH_URL", "http://localhost:9200")
        self.prometheus_url = os.getenv("PROMETHEUS_URL", "http://localhost:8000")

        # Initialize monitoring components
        self.system_monitor = SystemMonitor()

        # Use existing Prometheus monitor if available (lazy initialization)
        self.prometheus_monitor = None
        self._prometheus_init_attempted = False

    async def _query_elasticsearch(self, index: str, query: dict) -> dict:
        """Execute query against Elasticsearch."""
        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f"{self.elasticsearch_url}/{index}/_search",
                    json=query,
                    headers={"Content-Type": "application/json"},
                ) as response:
                    if response.status == 200:
                        return await response.json()
                    else:
                        logger.error(f"Elasticsearch query failed: {response.status}")
                        return {"hits": {"hits": []}}
        except Exception as e:
            logger.error(f"Error querying Elasticsearch: {e}")
            return {"hits": {"hits": []}}

    def _parse_time_range(self, time_range: str) -> str:
        """Convert time range string to Elasticsearch format."""
        time_map = {"5m": "now-5m", "15m": "now-15m", "1h": "now-1h", "24h": "now-24h"}
        return time_map.get(time_range, "now-15m")

    async def get_system_metrics(
        self, time_range: str = "15m", limit: int = 100
    ) -> List[dict]:
        """Get system performance metrics from Elasticsearch."""
        query = {
            "size": limit,
            "sort": [{"@timestamp": "desc"}],
            "query": {
                "bool": {
                    "filter": [
                        {
                            "range": {
                                "@timestamp": {
                                    "gte": self._parse_time_range(time_range),
                                    "lte": "now",
                                }
                            }
                        }
                    ]
                }
            },
            "_source": [
                "@timestamp",
                "cpu_usage",
                "memory_usage",
                "disk_usage",
                "active_cameras",
            ],
        }

        result = await self._query_elasticsearch("system_metrics", query)

        metrics = []
        for hit in result.get("hits", {}).get("hits", []):
            source = hit["_source"]
            metrics.append(
                {
                    "timestamp": source.get("@timestamp"),
                    "cpu_usage": source.get("cpu_usage", 0),
                    "memory_usage": source.get("memory_usage", 0),
                    "disk_usage": source.get("disk_usage", 0),
                    "active_cameras": source.get("active_cameras", 0),
                }
            )

        # If no Elasticsearch data, get current system metrics
        if not metrics:
            self.system_monitor.update()
            current_metrics = self.system_monitor.get_metrics()
            metrics.append(
                {
                    "timestamp": datetime.now().isoformat(),
                    "cpu_usage": current_metrics.get("cpu_usage", 0),
                    "memory_usage": current_metrics.get("memory_usage", 0),
                    "disk_usage": current_metrics.get("disk_usage", 0),
                    "active_cameras": current_metrics.get("active_cameras", 0),
                }
            )

        return list(reversed(metrics))  # Return chronological order

    async def get_camera_metrics(self) -> List[dict]:
        """Get current metrics for all cameras."""
        query = {
            "size": 0,
            "aggs": {
                "cameras": {
                    "terms": {"field": "camera_id.keyword", "size": 100},
                    "aggs": {
                        "latest_metrics": {
                            "top_hits": {
                                "sort": [{"@timestamp": {"order": "desc"}}],
                                "size": 1,
                                "_source": [
                                    "fps_actual",
                                    "fps_target",
                                    "latency_ms",
                                    "status",
                                    "last_detection",
                                ],
                            }
                        }
                    },
                }
            },
            "query": {"range": {"@timestamp": {"gte": "now-5m"}}},
        }

        result = await self._query_elasticsearch("detection-metrics-*", query)

        cameras = []
        for bucket in (
            result.get("aggregations", {}).get("cameras", {}).get("buckets", [])
        ):
            camera_id = bucket["key"]
            latest = bucket["latest_metrics"]["hits"]["hits"]

            if latest:
                source = latest[0]["_source"]
                cameras.append(
                    {
                        "camera_id": camera_id,
                        "fps_actual": source.get("fps_actual", 0),
                        "fps_target": source.get("fps_target", 30),
                        "latency_ms": source.get("latency_ms", 0),
                        "status": source.get("status", "unknown"),
                        "last_detection": source.get("last_detection"),
                    }
                )
            else:
                cameras.append(
                    {
                        "camera_id": camera_id,
                        "fps_actual": 0,
                        "fps_target": 30,
                        "latency_ms": 0,
                        "status": "offline",
                        "last_detection": None,
                    }
                )

        return cameras

    async def get_camera_performance(
        self, camera_id: str, time_range: str = "1h"
    ) -> dict:
        """Get detailed performance metrics for a specific camera."""
        query = {
            "size": 100,
            "sort": [{"@timestamp": "desc"}],
            "query": {
                "bool": {
                    "filter": [
                        {"term": {"camera_id.keyword": camera_id}},
                        {
                            "range": {
                                "@timestamp": {
                                    "gte": self._parse_time_range(time_range),
                                    "lte": "now",
                                }
                            }
                        },
                    ]
                }
            },
            "_source": [
                "@timestamp",
                "fps_actual",
                "latency_ms",
                "queue_depth",
                "dropped_frames",
            ],
        }

        result = await self._query_elasticsearch("detection-metrics-*", query)

        performance_data = []
        for hit in result.get("hits", {}).get("hits", []):
            source = hit["_source"]
            performance_data.append(
                {
                    "timestamp": source.get("@timestamp"),
                    "fps_actual": source.get("fps_actual", 0),
                    "latency_ms": source.get("latency_ms", 0),
                    "queue_depth": source.get("queue_depth", 0),
                    "dropped_frames": source.get("dropped_frames", 0),
                }
            )

        return {
            "camera_id": camera_id,
            "time_range": time_range,
            "data": list(reversed(performance_data)),
        }

    async def get_detection_metrics(
        self,
        time_range: str = "1h",
        camera_id: Optional[str] = None,
        confidence_threshold: float = 0.0,
    ) -> List[dict]:
        """Get detection metrics with optional filtering."""
        filters = [
            {
                "range": {
                    "@timestamp": {
                        "gte": self._parse_time_range(time_range),
                        "lte": "now",
                    }
                }
            },
            {"range": {"prediction.confidence": {"gte": confidence_threshold}}},
        ]

        if camera_id:
            filters.append({"term": {"camera_id.keyword": camera_id}})

        query = {
            "size": 100,
            "sort": [{"@timestamp": "desc"}],
            "query": {"bool": {"filter": filters}},
            "_source": [
                "@timestamp",
                "camera_id",
                "prediction.confidence",
                "prediction.label",
                "prediction.is_shoplifting",
                "alert.triggered",
            ],
        }

        result = await self._query_elasticsearch("detection-metrics-*", query)

        detections = []
        for hit in result.get("hits", {}).get("hits", []):
            source = hit["_source"]
            prediction = source.get("prediction", {})
            alert = source.get("alert", {})

            detections.append(
                {
                    "camera_id": source.get("camera_id", "unknown"),
                    "confidence": prediction.get("confidence", 0),
                    "label": prediction.get("label", "unknown"),
                    "is_shoplifting": prediction.get("is_shoplifting", False),
                    "timestamp": source.get("@timestamp"),
                    "alert_triggered": alert.get("triggered", False),
                }
            )

        return detections

    async def get_metrics_summary(self) -> dict:
        """Get comprehensive metrics summary."""
        # Get latest system metrics
        system_metrics = await self.get_system_metrics(time_range="5m", limit=1)
        latest_system = (
            system_metrics[0]
            if system_metrics
            else {
                "timestamp": datetime.now().isoformat(),
                "cpu_usage": 0,
                "memory_usage": 0,
                "disk_usage": 0,
                "active_cameras": 0,
            }
        )

        # Get camera metrics
        cameras = await self.get_camera_metrics()

        # Get recent detections
        recent_detections = await self.get_detection_metrics(
            time_range="1h", confidence_threshold=0.5
        )

        # Count today's detections and alerts
        today_detections = await self.get_detection_metrics(time_range="24h")
        today_alerts = [d for d in today_detections if d.get("alert_triggered", False)]

        return {
            "system": latest_system,
            "cameras": cameras,
            "recent_detections": recent_detections[:10],  # Last 10 detections
            "total_detections_today": len(today_detections),
            "alert_count_today": len(today_alerts),
        }

    def _ensure_prometheus_monitor(self):
        """Lazy initialization of Prometheus monitor to avoid timing issues."""
        if self._prometheus_init_attempted:
            return

        self._prometheus_init_attempted = True
        try:
            if get_monitor:
                self.prometheus_monitor = get_monitor()
                logger.info("Successfully connected to existing Prometheus monitor")
            elif PrometheusMonitor:
                self.prometheus_monitor = PrometheusMonitor()
                logger.info("Successfully initialized new Prometheus monitor")
            else:
                logger.info("Prometheus monitoring not available")
        except Exception as e:
            logger.debug(f"Prometheus monitor not yet available: {e}")
            # Reset flag to try again later
            self._prometheus_init_attempted = False

    async def get_health_status(self) -> dict:
        """Check health of metrics infrastructure."""
        # Try to initialize Prometheus monitor if not done yet
        self._ensure_prometheus_monitor()

        health = {
            "elasticsearch": False,
            "prometheus": False,
            "system_monitor": True,
            "timestamp": datetime.now().isoformat(),
        }

        # Check Elasticsearch
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.elasticsearch_url}/_cluster/health"
                ) as response:
                    if response.status == 200:
                        health["elasticsearch"] = True
        except Exception as e:
            logger.error(f"Elasticsearch health check failed: {e}")

        # Check Prometheus
        if self.prometheus_monitor:
            health["prometheus"] = True

        return health

    async def get_recent_alerts(
        self, limit: int = 50, severity: Optional[str] = None
    ) -> List[dict]:
        """Get recent alerts from the system."""
        filters = [
            {"term": {"alert.triggered": True}},
            {"range": {"@timestamp": {"gte": "now-24h"}}},
        ]

        if severity:
            filters.append({"term": {"alert.level.keyword": severity}})

        query = {
            "size": limit,
            "sort": [{"@timestamp": "desc"}],
            "query": {"bool": {"filter": filters}},
            "_source": [
                "@timestamp",
                "camera_id",
                "alert.level",
                "alert.type",
                "prediction.confidence",
                "prediction.label",
            ],
        }

        result = await self._query_elasticsearch("detection-metrics-*", query)

        alerts = []
        for hit in result.get("hits", {}).get("hits", []):
            source = hit["_source"]
            alert = source.get("alert", {})
            prediction = source.get("prediction", {})

            alerts.append(
                {
                    "timestamp": source.get("@timestamp"),
                    "camera_id": source.get("camera_id", "unknown"),
                    "level": alert.get("level", "unknown"),
                    "type": alert.get("type", "detection"),
                    "confidence": prediction.get("confidence", 0),
                    "label": prediction.get("label", "unknown"),
                }
            )

        return alerts
